File: kotlin/bc-envelope/src/main/kotlin/com/blockchaincommons/bcenvelope/Request.kt
package com.blockchaincommons.bcenvelope

import com.blockchaincommons.bccomponents.ARID
import com.blockchaincommons.bctags.TAG_REQUEST
import com.blockchaincommons.dcbor.Cbor
import com.blockchaincommons.dcbor.CborDate
import com.blockchaincommons.dcbor.tagsForValues
import com.blockchaincommons.knownvalues.BODY
import com.blockchaincommons.knownvalues.DATE
import com.blockchaincommons.knownvalues.NOTE

/**
 * A request message for executing a function with parameters.
 *
 * Requests are part of the expression system that enables distributed function
 * calls. Each request contains a body (an [Expression]) and a unique [ARID]
 * identifier for tracking and correlation.
 */
class Request private constructor(
    @PublishedApi internal var body: Expression,
    val id: ARID,
    private val note: String = "",
    private val date: CborDate? = null,
) {
    /** Creates a new request from an expression body and ID. */
    constructor(body: Expression, id: ARID) : this(
        body = body,
        id = id,
        note = "",
        date = null,
    )

    /** Creates a new request with a function and ID. */
    constructor(function: Function, id: ARID) : this(Expression(function), id)

    /** Creates a new request with a named function and ID. */
    constructor(name: String, id: ARID) : this(Expression(name), id)

    /** Returns the body expression of the request. */
    fun body(): Expression = body

    /** Returns the unique identifier of the request. */
    fun id(): ARID = id

    /** Returns the note attached to the request, or empty string. */
    fun note(): String = note

    /** Returns the date attached to the request, if any. */
    fun date(): CborDate? = date

    /** Returns a new request with the given note. */
    fun withNote(newNote: String): Request =
        Request(body, id, newNote, date)

    /** Returns a new request with the given date. */
    fun withDate(newDate: CborDate): Request =
        Request(body, id, note, newDate)

    /** Adds a parameter to the request. */
    fun withParameter(parameter: Parameter, value: Any): Request {
        body = body.withParameter(parameter, value)
        return this
    }

    /** Adds an optional parameter to the request. */
    fun withOptionalParameter(parameter: Parameter, value: Any?): Request {
        body = body.withOptionalParameter(parameter, value)
        return this
    }

    /** Returns the argument for the given parameter. */
    fun objectForParameter(param: Parameter): Envelope =
        body.objectForParameter(param)

    /** Returns all arguments for the given parameter. */
    fun objectsForParameter(param: Parameter): List<Envelope> =
        body.objectsForParameter(param)

    /** Returns the argument for the given parameter, decoded as a specific type. */
    inline fun <reified T : Any> extractObjectForParameter(param: Parameter): T =
        body.extractObjectForParameter(param)

    /** Returns the argument for the given parameter as the given type, or null. */
    inline fun <reified T : Any> extractOptionalObjectForParameter(param: Parameter): T? =
        body.extractOptionalObjectForParameter(param)

    /** Returns the function of the request. */
    fun function(): Function = body.function()

    /** Returns the expression envelope of the request. */
    fun expressionEnvelope(): Envelope = body.expressionEnvelope()

    /** Returns a human-readable summary. */
    fun summary(): String =
        "id: ${id.shortDescription()}, body: ${body.expressionEnvelope().formatFlat()}"

    /** Converts this request to an envelope. */
    fun toEnvelope(): Envelope {
        val requestTag = tagsForValues(listOf(TAG_REQUEST)).first()
        var envelope = Envelope.from(Cbor.tagged(requestTag, id.taggedCbor()))
            .addAssertion(BODY, body.toEnvelope())
        if (note.isNotEmpty()) {
            envelope = envelope.addAssertion(NOTE, note)
        }
        val d = date
        if (d != null) {
            envelope = envelope.addAssertion(DATE, d)
        }
        return envelope
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Request) return false
        return id == other.id && body == other.body && note == other.note && date == other.date
    }

    override fun hashCode(): Int {
        var result = body.hashCode()
        result = 31 * result + id.hashCode()
        result = 31 * result + note.hashCode()
        result = 31 * result + (date?.hashCode() ?: 0)
        return result
    }

    override fun toString(): String = "Request(${summary()})"

    companion object {
        /** Parses a request from an envelope. */
        fun fromEnvelope(
            envelope: Envelope,
            expectedFunction: Function? = null,
        ): Request {
            val bodyEnvelope = envelope.objectForPredicate(BODY)
            val expression = Expression.fromEnvelope(bodyEnvelope, expectedFunction)

            val requestTag = tagsForValues(listOf(TAG_REQUEST)).first()
            val (tag, idCbor) = envelope.subject().tryLeaf().tryTagged()
            if (tag != requestTag) {
                throw EnvelopeException.InvalidFormat()
            }
            val id = ARID.fromTaggedCbor(idCbor)

            val note: String = envelope.extractOptionalObjectForPredicate(NOTE) ?: ""
            val date: CborDate? = envelope.extractOptionalObjectForPredicate<CborDate>(DATE)

            return Request(expression, id, note, date)
        }
    }
}

File: csharp/BCEnvelope/BCEnvelope/Request.cs
using BlockchainCommons.BCComponents;
using BlockchainCommons.BCTags;
using BlockchainCommons.DCbor;
using BlockchainCommons.KnownValues;

namespace BlockchainCommons.BCEnvelope;

/// <summary>
/// A request message for executing a function with parameters.
/// </summary>
/// <remarks>
/// <para>
/// Requests are part of the expression system that enables distributed function
/// calls and communication between systems. Each request contains a body (an
/// <see cref="Expression"/>) that represents the function to be executed, and
/// a unique identifier (<see cref="ARID"/>) for tracking and correlation.
/// </para>
/// <para>
/// When serialized to an envelope, requests are tagged with #6.40004 (TAG_REQUEST).
/// </para>
/// </remarks>
public sealed class Request
{
    private Expression _body;
    private readonly ARID _id;
    private readonly string _note;
    private readonly CborDate? _date;

    private Request(Expression body, ARID id, string note, CborDate? date)
    {
        _body = body;
        _id = id;
        _note = note;
        _date = date;
    }

    /// <summary>
    /// Creates a new request from an expression body and ID.
    /// </summary>
    /// <param name="body">The expression to be executed.</param>
    /// <param name="id">Unique identifier for the request.</param>
    public Request(Expression body, ARID id)
        : this(body, id, "", null)
    {
    }

    /// <summary>
    /// Creates a new request with a function and ID.
    /// </summary>
    /// <param name="function">The function to be executed.</param>
    /// <param name="id">Unique identifier for the request.</param>
    public Request(Function function, ARID id)
        : this(new Expression(function), id)
    {
    }

    /// <summary>
    /// Creates a new request with a named function and ID.
    /// </summary>
    /// <param name="name">The function name string.</param>
    /// <param name="id">Unique identifier for the request.</param>
    public Request(string name, ARID id)
        : this(new Expression(name), id)
    {
    }

    /// <summary>Returns the body expression of the request.</summary>
    public Expression Body => _body;

    /// <summary>Returns the unique identifier of the request.</summary>
    public ARID Id => _id;

    /// <summary>Returns the note attached to the request, or empty string.</summary>
    public string Note => _note;

    /// <summary>Returns the date attached to the request, if any.</summary>
    public CborDate? Date => _date;

    /// <summary>Returns the function of the request.</summary>
    public Function Function => _body.Function;

    /// <summary>Returns the expression envelope of the request.</summary>
    public Envelope ExpressionEnvelope => _body.ExpressionEnvelope;

    /// <summary>
    /// Returns a new request with the given note.
    /// </summary>
    /// <param name="note">The note to attach.</param>
    /// <returns>A new <see cref="Request"/> with the note.</returns>
    public Request WithNote(string note) =>
        new(_body, _id, note, _date);

    /// <summary>
    /// Returns a new request with the given date.
    /// </summary>
    /// <param name="date">The date to attach.</param>
    /// <returns>A new <see cref="Request"/> with the date.</returns>
    public Request WithDate(CborDate date) =>
        new(_body, _id, _note, date);

    /// <summary>
    /// Adds a parameter to the request.
    /// </summary>
    /// <param name="parameter">The parameter identifier.</param>
    /// <param name="value">The value for the parameter.</param>
    /// <returns>This request with the parameter added.</returns>
    public Request WithParameter(Parameter parameter, object value)
    {
        _body = _body.WithParameter(parameter, value);
        return this;
    }

    /// <summary>
    /// Adds an optional parameter to the request.
    /// </summary>
    /// <param name="parameter">The parameter identifier.</param>
    /// <param name="value">The optional value for the parameter.</param>
    /// <returns>This request, with the parameter added if the value is not null.</returns>
    public Request WithOptionalParameter(Parameter parameter, object? value)
    {
        _body = _body.WithOptionalParameter(parameter, value);
        return this;
    }

    /// <summary>
    /// Returns the argument for the given parameter.
    /// </summary>
    public Envelope ObjectForParameter(Parameter param) =>
        _body.ObjectForParameter(param);

    /// <summary>
    /// Returns all arguments for the given parameter.
    /// </summary>
    public List<Envelope> ObjectsForParameter(Parameter param) =>
        _body.ObjectsForParameter(param);

    /// <summary>
    /// Returns the argument for the given parameter, decoded as the specified type.
    /// </summary>
    public T ExtractObjectForParameter<T>(Parameter param) =>
        _body.ExtractObjectForParameter<T>(param);

    /// <summary>
    /// Returns the argument for the given parameter as the given type, or default.
    /// </summary>
    public T? ExtractOptionalObjectForParameter<T>(Parameter param) =>
        _body.ExtractOptionalObjectForParameter<T>(param);

    /// <summary>
    /// Returns a human-readable summary.
    /// </summary>
    public string Summary() =>
        $"id: {_id.ShortDescription()}, body: {_body.ExpressionEnvelope}";

    /// <summary>
    /// Converts this request to an envelope.
    /// </summary>
    public Envelope ToEnvelope()
    {
        var requestTag = GlobalTags.TagsForValues(BcTags.TagRequest)[0];
        var envelope = Envelope.Create(Cbor.ToTaggedValue(requestTag, _id.TaggedCbor()))
            .AddAssertion(KnownValuesRegistry.Body, _body.ToEnvelope())
            .AddAssertionIf(!string.IsNullOrEmpty(_note), KnownValuesRegistry.Note, _note)
            .AddOptionalAssertion(KnownValuesRegistry.Date, _date);
        return envelope;
    }

    /// <summary>
    /// Parses a request from an envelope.
    /// </summary>
    /// <param name="envelope">The envelope to parse.</param>
    /// <param name="expectedFunction">An optional expected function for validation.</param>
    /// <returns>A new <see cref="Request"/>.</returns>
    public static Request FromEnvelope(Envelope envelope, Function? expectedFunction = null)
    {
        var bodyEnvelope = envelope.ObjectForPredicate(KnownValuesRegistry.Body);
        var expression = Expression.FromEnvelope(bodyEnvelope, expectedFunction);

        var requestTag = GlobalTags.TagsForValues(BcTags.TagRequest)[0];
        var idCbor = envelope.Subject.TryLeaf().TryIntoExpectedTaggedValue(requestTag);
        var id = ARID.FromTaggedCbor(idCbor);

        var note = envelope.ExtractObjectForPredicateWithDefault<string>(
            KnownValuesRegistry.Note, "");
        var date = envelope.ExtractOptionalObjectForPredicate<CborDate>(
            KnownValuesRegistry.Date);

        return new Request(expression, id, note, date);
    }

    /// <inheritdoc/>
    public override bool Equals(object? obj)
    {
        if (obj is not Request other) return false;
        if (ReferenceEquals(this, other)) return true;
        return _id == other._id
            && _body.Equals(other._body)
            && _note == other._note
            && Equals(_date, other._date);
    }

    /// <inheritdoc/>
    public override int GetHashCode() =>
        HashCode.Combine(_body, _id, _note, _date);

    /// <inheritdoc/>
    public override string ToString() => $"Request({Summary()})";
}
